connmgr/connmanager.go

```diff
@@ -0,0 +1,235 @@
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package connmgr
+
+import (
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// DialFunc defines a function that dials a connection.
+type DialFunc func(string, string) (net.Conn, error)
+
+// AddressFunc defines a function that returns a network address to connect to.
+type AddressFunc func() string
+
+// ConnState represents the state of the requested connection.
+type ConnState uint8
+
+// ConnState can be either pending, established, disconnected or failed.  When
+// a new connection is requested, it is attempted and categorized as
+// established or failed depending on the connection result.  An established
+// connection which was disconnected is categorized as disconnected.
+const (
+	ConnPending      ConnState = 0x0
+	ConnEstablished  ConnState = 0x1
+	ConnDisconnected ConnState = 0x2
+	ConnFailed       ConnState = 0x3
+)
+
+// ConnReq is the connection request to a network address. If permanent, the
+// connection will be retried on disconnection.
+type ConnReq struct {
+	Addr      string
+	Permanent bool
+
+	state      ConnState
+	retryCount uint32
+}
+
+// State is the connection state of the requested connection.
+func (c *ConnReq) State() ConnState {
+	return c.state
+}
+
+// ConnectedFunc is the signature of the callback function which is used to
+// subscribe to new connections.
+type ConnectedFunc func(*ConnReq, net.Conn)
+
+const (
+	// maxRetryDuration is the max amount of time retrying of a persistent
+	// connection is allowed to grow to.  This is necessary since the retry
+	// logic uses a backoff mechanism which increases the interval base done
+	// the number of retries that have been done.
+	maxRetryDuration = time.Minute * 5
+)
+
+// Config holds the configuration options related to the connection manager.
+type Config struct {
+	// MaxOutboundConnections is the maximum number of outbound network
+	// connections to maintain.
+	MaxOutboundConnections uint32
+
+	// RetryDuration is the duration to wait before retrying a permanent
+	// connection request.
+	RetryDuration time.Duration
+
+	// OnConnection is a callback that is fired when a new connection is established.
+	OnConnection ConnectedFunc
+
+	// GetNewAddress is a way to get an address to make a network connection to.
+	GetNewAddress AddressFunc
+
+	// Dial connects to the address on the named network.
+	Dial DialFunc
+}
+
+// handleConnected is used to queue a successful connection.
+type handleConnected struct {
+	c    *ConnReq
+	conn net.Conn
+}
+
+// handleDisconnected is used to remove a connection.
+type handleDisconnected struct {
+	c *ConnReq
+}
+
+// handleFailed is used to remove a pending connection.
+type handleFailed struct {
+	c   *ConnReq
+	err error
+}
+
+// ConnManager provides a manager to handle network connections.
+type ConnManager struct {
+	cfg      *Config
+	wg       sync.WaitGroup
+	conns    map[*ConnReq]net.Conn
```

Comment:
Using points in a map is dangerous because it's too easy to end up with a pointer to a copy of the object which is otherwise functionally the same.  I would suggest giving a `ConnReq` a unique uint64 id (probably atomically incremented on the `ConnManager` instance itself) and keying the map by uint64.